dulwich/porcelain.py

```diff
@@ -1859,6 +1863,114 @@ def reset_file(repo, file_path: str, target: bytes = b'HEAD',
 
 
+def _update_head_during_checkout_branch(repo, target):
+    checkout_target = None
+    if target == b'HEAD':  # Do not update head while trying to checkout to HEAD.
+        pass
+    elif target in repo.refs.keys(base=LOCAL_BRANCH_PREFIX):
+        update_head(repo, target)
+    else:
+        # If checking out a remote branch, create a local one without the remote name prefix.
+        config = repo.get_config()
+        name = target.split(b"/")[0]
+        section = (b"remote", name)
+        if config.has_section(section):
+            checkout_target = target.replace(name + b"/", b"")
+            try:
+                branch_create(repo, checkout_target, (LOCAL_REMOTE_PREFIX + target).decode())
+            except Error:
+                pass
+            update_head(repo, LOCAL_BRANCH_PREFIX + checkout_target)
+        else:
+            update_head(repo, target, detached=True)
+
+    return checkout_target
+
+
+def checkout_branch(repo, target: Union[bytes, str], force: bool = False):
+    """switch branches or restore working tree files
+    Args:
+      repo: dulwich Repo object
+      target: branch name or commit sha to checkout
+      force: true or not to force checkout
+    """
+    target = to_bytes(target)
+
+    current_tree = parse_tree(repo, repo.head())
+    target_tree = parse_tree(repo, target)
+
+    if force:
+        repo.reset_index(target_tree.id)
+        _update_head_during_checkout_branch(repo, target)
+    else:
+        status_report = status(repo)
+        changes = list(set(status_report[0]['add'] + status_report[0]['delete'] + status_report[0]['modify'] + status_report[1]))
+        index = 0
+        while index < len(changes):
+            change = changes[index]
+            try:
+                current_tree.lookup_path(repo.object_store.__getitem__, change)
+                try:
+                    target_tree.lookup_path(repo.object_store.__getitem__, change)
+                    index += 1
+                except KeyError:
+                    raise AbortCheckout('Your local changes to the following files would be overwritten by checkout: ' + change.decode())
+            except KeyError:
+                changes.pop(index)
+
+        # Update head.
+        checkout_target = _update_head_during_checkout_branch(repo, target)
+        if checkout_target is not None:
+            target_tree = parse_tree(repo, checkout_target)
+
+        dealt_with = []
+        repo_index = repo.open_index()
+        for entry in iter_tree_contents(repo.object_store, target_tree.id):
+            dealt_with.append(entry.path)
+            if entry.path in changes:
+                continue
+            full_path = os.path.join(os.fsencode(repo.path), entry.path)
+            blob = repo.object_store[entry.sha]
+            ensure_dir_exists(os.path.dirname(full_path))
+            st = build_file_from_blob(blob, entry.mode, full_path)
+            st_tuple = (
```

Comment:
There should be no need to construct st_tuple here, you can just pass "st" to index_entry_from_stat. if necessary, you can pass a custom mode to index_entry_from_stat.

Reply:
[1d5b8d4](https://github.com/jelmer/dulwich/pull/1148/commits/1d5b8d42cb8b679561daf5aae71b08e277d426cf)